src/Tools/BuildValidator/CompilationDiff.cs

```diff
@@ -114,124 +130,183 @@ private CompilationDiff(ImmutableArray<Diagnostic> diagnostics, string originalP
-                            | PdbToXmlOptions.ThrowOnError
-                            | PdbToXmlOptions.ExcludeScopes
-                            | PdbToXmlOptions.IncludeSourceServerInformation
-                            | PdbToXmlOptions.IncludeEmbeddedSources
-                            | PdbToXmlOptions.IncludeTokens
-                            | PdbToXmlOptions.IncludeMethodSpans;
+            static BuildDataFiles createBuildArtifacts(string outputPath, string assemblyFileName, BuildInfo buildInfo)
+            {
+                var assemblyName = Path.GetFileNameWithoutExtension(assemblyFileName);
+                var assemblyFilePath = Path.Combine(outputPath, assemblyFileName);
+                var buildDataFiles = new BuildDataFiles(
+                    AssemblyMdvFilePath: Path.Combine(outputPath, assemblyName + ".mdv"),
+                    PdbMdvFilePath: Path.Combine(outputPath, assemblyName + ".pdb.mdv"),
+                    ILFilePath: Path.Combine(outputPath, assemblyName + ".il"),
+                    PdbXmlFilePath: Path.Combine(outputPath, assemblyName + ".pdb.xml"),
+                    CustomDataFilePath: Path.Combine(outputPath, "custom-data.txt"));
+
+                Directory.CreateDirectory(outputPath);
+                File.WriteAllBytes(assemblyFilePath, buildInfo.AssemblyBytes);
+
+                // This is deliberately named .extracted.pdb instead of .pdb. A number of tools will look
+                // for a PDB with the name assemblyName.pdb. Want to make explicitly sure that does not 
+                // happen and such tools always correctly fall back to the embedded PDB. 
+                var pdbFilePath = Path.Combine(outputPath, assemblyName + ".extracted.pdb");
+                writeAllBytes(pdbFilePath, new Span<byte>(buildInfo.PdbMetadataReader.MetadataPointer, buildInfo.PdbMetadataReader.MetadataLength));
+
+                createMetadataVisualization(buildDataFiles.AssemblyMdvFilePath, buildInfo.AssemblyMetadataReader);
+                createMetadataVisualization(buildDataFiles.PdbMdvFilePath, buildInfo.PdbMetadataReader);
+                createDataFile(buildDataFiles.CustomDataFilePath, buildInfo.AssemblyReader, buildInfo.PdbMetadataReader);
+
+                var pdbToXmlOptions = PdbToXmlOptions.ResolveTokens
+                    | PdbToXmlOptions.ThrowOnError
+                    | PdbToXmlOptions.ExcludeScopes
+                    | PdbToXmlOptions.IncludeSourceServerInformation
+                    | PdbToXmlOptions.IncludeEmbeddedSources
+                    | PdbToXmlOptions.IncludeTokens
+                    | PdbToXmlOptions.IncludeMethodSpans;
+
+                using var pdbXmlStream = File.Create(buildDataFiles.PdbXmlFilePath);
+                PdbToXmlConverter.ToXml(
+                    new StreamWriter(pdbXmlStream),
+                    pdbStream: new UnmanagedMemoryStream(buildInfo.PdbMetadataReader.MetadataPointer, buildInfo.PdbMetadataReader.MetadataLength),
```

Comment:
>new UnmanagedMemoryStream [](start = 31, length = 25)

This wasn't introduced by your change but ... should we `Dispose()` this instance after calling `ToXml()`?

Reply:
It's like `MemoryStream` in that it has a no-op dispose and no finalizer so it's not necessary. When easy I toss it into a `using` but that's more pedantic behavior than anything else.